modules/indexing/src/test/java/org/apache/ignite/internal/metric/IndexPagesMetricsPageDisplacementTest.java

```diff
@@ -0,0 +1,189 @@
+     * go down.
+     */
+    @Test
+    public void testPageDisplacement() throws IgniteCheckedException {
+        int grpId = grid.cachex(TEST_CACHE_NAME).context().groupId();
+
+        DataRegion dataRegion = grid.context().cache().context().database().dataRegion(null);
+        PageMetrics pageMetrics = dataRegion.metrics().cacheGrpPageMetrics(grpId);
+
+        int personId = 0;
+        long idxPagesOnDisk;
+        long idxPagesInMemory;
+
+        do {
+            // insert data into the cache until some index pages get displaced to the storage
+            for (int i = 0; i < 100; i++, personId++)
+                cache.put(personId, new Person(personId, "foobar"));
+
+            idxPagesOnDisk = getIdxPagesOnDisk(grpId).size();
+            idxPagesInMemory = idxPageCounter.countIdxPagesInMemory(grpId);
+
+            assertThat(pageMetrics.indexPages().value(), is(idxPagesInMemory));
+
+        } while (idxPagesOnDisk <= idxPagesInMemory);
+
+        // load pages back into memory and check that the metric value has increased
+        touchIdxPages(grpId);
+
+        long allIdxPagesInMemory = idxPageCounter.countIdxPagesInMemory(grpId);
+
+        assertThat(allIdxPagesInMemory, greaterThan(idxPagesInMemory));
+        assertThat(pageMetrics.indexPages().value(), is(allIdxPagesInMemory));
+    }
+
+    /**
+     * Acquires all pages inside the index partition for the given cache group in order to load them back into
+     * memory if some of them had been displaced to the storage earlier.
+     */
+    private void touchIdxPages(int grpId) throws IgniteCheckedException {
+        DataRegion dataRegion = grid.context().cache().context().database().dataRegion(null);
+        PageMemory pageMemory = dataRegion.pageMemory();
+
+        for (long pageId : getIdxPagesOnDisk(grpId))
+            // acquire, but not release all pages, so that they don't get displaced back to the storage
+            pageMemory.acquirePage(grpId, pageId);
+    }
+
+    /**
+     * Returns IDs of index pages currently residing in the storage.
+     */
+    private List<Long> getIdxPagesOnDisk(int grpId) throws IgniteCheckedException {
+        FilePageStoreManager pageStoreMgr = (FilePageStoreManager) grid.context().cache().context().pageStore();
+        FilePageStore pageStore = (FilePageStore) pageStoreMgr.getStore(grpId, PageIdAllocator.INDEX_PARTITION);
+
+        List<Long> result = new ArrayList<>();
+
+        ByteBuffer buf = ByteBuffer
+            .allocateDirect(pageStore.getPageSize())
+            .order(ByteOrder.nativeOrder());
+
+        // Page Store contains a one-page header
+        long numPages = pageStore.size() / pageStore.getPageSize() - 1;
+
+        for (int i = 0; i < numPages; i++) {
+            long pageId = PageIdUtils.pageId(PageIdAllocator.INDEX_PARTITION, (byte)0, i);
+
+            try {
+                pageStore.read(pageId, buf, false);
+            } catch (IgniteDataIntegrityViolationException ignored) {
```

Comment:
Is this the exception that you don't have locally?

Reply:
yes

Reply:
I forgot to add a comment here, will do